source/common/ssl/context_impl.cc

```diff
@@ -812,11 +820,125 @@ int ServerContextImpl::sessionTicketProcess(SSL*, uint8_t* key_name, uint8_t* iv
   }
 }
 
+bool ServerContextImpl::isClientEcdsaCapable(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  CBS client_hello;
+  CBS_init(&client_hello, ssl_client_hello->client_hello, ssl_client_hello->client_hello_len);
+
+  uint16_t client_version;
+  if (!CBS_get_u16(&client_hello, &client_version)) {
+    return false;
+  }
+
+  // This is the TLSv1.3 case (TLSv1.2 on the wire and the supported_versions extensions present).
+  // We just need to loook at signature algorithms.
+  if (client_version == TLS1_2_VERSION && tls_max_version_ == TLS1_3_VERSION) {
+    // If the supported_versions extension is found then we assume that the client is competent
+    // enough that just checking the signature_algorithms is sufficient.
+    const uint8_t* supported_versions_data;
+    size_t supported_versions_len;
+    if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_supported_versions,
+                                             &supported_versions_data, &supported_versions_len)) {
+      const uint8_t* signature_algorithms_data;
+      size_t signature_algorithms_len;
+      if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_signature_algorithms,
+                                               &signature_algorithms_data,
+                                               &signature_algorithms_len)) {
+        CBS signature_algorithms_ext, signature_algorithms;
+        CBS_init(&signature_algorithms_ext, signature_algorithms_data, signature_algorithms_len);
+        if (!CBS_get_u16_length_prefixed(&signature_algorithms_ext, &signature_algorithms) ||
+            CBS_len(&signature_algorithms_ext) != 0) {
+          return false;
+        }
+
+        while (CBS_len(&signature_algorithms) > 0) {
+          uint16_t sigalg;
+          if (!CBS_get_u16(&signature_algorithms, &sigalg)) {
+            return false;
+          }
+
+          if (sigalg == SSL_SIGN_ECDSA_SECP256R1_SHA256) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+  }
+
+  // Otherwise we are < TLSv1.3 and need to look at both the curves in the supported_groups for
+  // ECDSA and also for a compatible cipher suite. https://tools.ietf.org/html/rfc4492#section-5.1.1
+  const uint8_t* curvelist_data;
+  size_t curvelist_len;
+  if (!SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_supported_groups,
+                                            &curvelist_data, &curvelist_len)) {
+    return false;
+  }
+
+  CBS curvelist;
+  CBS_init(&curvelist, curvelist_data, curvelist_len);
+
+  bool p256_ok = false;
+  while (CBS_len(&curvelist) > 0) {
+    uint16_t named_curve;
+    if (!CBS_get_u16(&curvelist, &named_curve)) {
+      return false;
+    }
+
+    if (named_curve == 23 /* secp256r1 */) {
```

Comment:
Please check for stronger curves as well, i.e.
```
if (named_curve  == SSL_CURVE_SECP256R1 || named_curve  == SSL_CURVE_SECP384R1 ||
    named_curve  == SSL_CURVE_SECP521R1) {
```

Reply:
That's not quite right either. You need to specifically check for the curve that your certificate uses. The code this was based on assumes server ECDSA certificates are always P-256. (P-256 receives by *far* the bulk of security hardening and performance work. Until earlier this year, our P-384 and P-521 implementations were not even constant-time, and we are unlikely to pay the binary size to optimize either with pre-computed base point tables.)

If Envoy also only cares about P-256 (good idea) this code should stay as-is (but do replace 23 with `SSL_CURVE_SECP256R1`). If it wants to support others, it should extract the curve type and look for just that one.

Reply:
I'm going to follow the logic that will arrive with SSL_IDENTITY. I'll find the first ECDSA certificate in the list, extract its curve and use this in the check. Does this make sense to you and @davidben?

Reply:
Yeah.

Reply:
If you extract just the first ECDSA certificate's curve, then you should only consider the first certificate because the check won't be valid for the others.

But I would recommend just doing P-256. Again, the other curves do not see security or performance work.